### PR TITLE
Create solver class from reasoning based simplifier

### PR DESCRIPTION
**Pull request by hrkrshnn**
Extracted parts of https://github.com/ethereum/solidity/pull/11192

### CODE REVIEW

File: libyul/optimiser/ReasoningBasedSimplifier.cpp [hunk 1]
```diff
@@ -56,14 +56,8 @@ std::optional<string> ReasoningBasedSimplifier::invalidInCurrentEnvironment()
...
 {
-	if (_varDecl.variables.size() != 1 || !_varDecl.value)
-		return;
-	YulString varName = _varDecl.variables.front().name;
-	if (!m_ssaVariables.count(varName))
-		return;
-	bool const inserted = m_variables.insert({varName, m_solver->newVariable("yul_" + varName.str(), defaultSort())}).second;
-	yulAssert(inserted, "");
-	m_solver->addAssertion(m_variables.at(varName) == encodeExpression(*_varDecl.value));
+	ASTModifier::operator()(_varDecl);
```
Comment (hrkrshnn) [inline libyul/optimiser/ReasoningBasedSimplifier.cpp]: Visit to `_varDecl.value` was missing before.

Comment (chriseth) [reply libyul/optimiser/ReasoningBasedSimplifier.cpp]: Why does it need the visit?

File: libyul/optimiser/Solver.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+class Expression;
+struct Sort;
+
+}
+
+namespace solidity::yul
+{
+
+/**
+ * A base class for SMT based optimization steps
```
Comment (chriseth) [inline libyul/optimiser/Solver.h]:
```suggestion
 * Base class for SMT based optimization steps
```

File: libyul/optimiser/Solver.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ */
+class Solver
+{
+protected:
+	/// A helper function that encodes VariableDeclaration
```
Comment (chriseth) [inline libyul/optimiser/Solver.h]:
```suggestion
	/// Helper function that encodes VariableDeclaration
```

File: libyul/optimiser/Solver.h (hunk 2 above)
Comment (chriseth) [reply libyul/optimiser/Solver.h]: Does it work best in SSA form?

File: libyul/optimiser/Solver.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	void encodeVariableDeclaration(VariableDeclaration const& _varDecl);
+
+	Solver(
```
Comment (chriseth) [inline libyul/optimiser/Solver.h]: Shouldn't the constructor go first?

File: libyul/optimiser/Solver.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	);
+
+	/// The encoding for a builtin. The type of encoding determines what we are
+	/// solving for.
+	virtual smtutil::Expression encodeEVMBuiltin(
+		evmasm::Instruction _instruction,
+		std::vector<Expression> const& _arguments
+	) = 0;
+
+	smtutil::Expression encodeExpression(
```
Comment (chriseth) [inline libyul/optimiser/Solver.h]: single line?

File: libyul/optimiser/Solver.h [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		Expression const& _expression
+	);
+
+	smtutil::Expression int2bv(smtutil::Expression _arg) const;
+	smtutil::Expression bv2int(smtutil::Expression _arg) const;
+
+	smtutil::Expression newVariable();
+	virtual smtutil::Expression newRestrictedVariable(bigint _maxValue=(bigint(1) << 256) - 1);
```
Comment (chriseth) [inline libyul/optimiser/Solver.h]:
```suggestion
	virtual smtutil::Expression newRestrictedVariable(bigint _maxValue = (bigint(1) << 256) - 1);
```

File: libyul/optimiser/Solver.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+smtutil::Expression Solver::newVariable()
+{
+	return m_solver->newVariable(uniqueName(), defaultSort());
+}
+
+smtutil::Expression Solver::newRestrictedVariable(bigint _maxValue)
+{
+	smtutil::Expression var = newVariable();
+	m_solver->addAssertion(0 <= var && var < smtutil::Expression(_maxValue));
```
Comment (chriseth) [inline libyul/optimiser/Solver.cpp]:
```suggestion
	m_solver->addAssertion(0 <= var && var <= smtutil::Expression(_maxValue));
```

File: libyul/optimiser/Solver.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+{
+	return _value;
+}
+
+smtutil::Expression Solver::literalValue(Literal const& _literal) const
+{
+	return smtutil::Expression(valueOfLiteral(_literal));
+}
+
+smtutil::Expression Solver::unsignedToSigned(smtutil::Expression _value)
```
Comment (chriseth) [inline libyul/optimiser/Solver.cpp]:
```suggestion
smtutil::Expression Solver::twosComplementToSigned(smtutil::Expression _value)
```

File: libyul/optimiser/Solver.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+{
+	return smtutil::Expression::ite(
+		_value < smtutil::Expression(bigint(1) << 255),
+		_value,
+		_value - smtutil::Expression(bigint(1) << 256)
+	);
+}
+
+smtutil::Expression Solver::signedToUnsigned(smtutil::Expression _value)
```
Comment (chriseth) [inline libyul/optimiser/Solver.cpp]:
```suggestion
smtutil::Expression Solver::signedToTwosComplement(smtutil::Expression _value)
```

File: libyul/optimiser/Solver.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+		_varDecl.variables.size() == 1 &&
+		_varDecl.value &&
+		m_ssaVariables.count(_varDecl.variables.front().name)
+	)
+	{
+		YulString const& variableName = _varDecl.variables.front().name;
+		bool const inserted = m_variables.insert({
+			variableName,
+			m_solver->newVariable("yul_" + variableName.str(), defaultSort())
+			}).second;
```
Comment (chriseth) [inline libyul/optimiser/Solver.cpp]:
```suggestion
		}).second;
```

File: libyul/optimiser/ReasoningBasedSimplifier.cpp (hunk 1 above)
Comment (hrkrshnn) [reply libyul/optimiser/ReasoningBasedSimplifier.cpp]: Actually, it doesn't need it for ReasoningBasedSimplifier. Should I remove it?

Comment (chriseth) [reply libyul/optimiser/ReasoningBasedSimplifier.cpp]: Yes, please remove it.

Comment (hrkrshnn) [inline libyul/optimiser/ReasoningBasedSimplifier.cpp]:
```suggestion
```

File: libyul/optimiser/Solver.h (hunk 2 above)
Comment (hrkrshnn) [reply libyul/optimiser/Solver.h]: Yes, I'll add a comment.

File: libyul/optimiser/ReasoningBasedSimplifier.cpp [hunk 11]
```diff
@@ -56,14 +56,7 @@ std::optional<string> ReasoningBasedSimplifier::invalidInCurrentEnvironment()
 
 void ReasoningBasedSimplifier::operator()(VariableDeclaration& _varDecl)
 {
-	if (_varDecl.variables.size() != 1 || !_varDecl.value)
```
Comment (chriseth) [inline libyul/optimiser/ReasoningBasedSimplifier.cpp:59]: Why is this change not just part of the first commit?

Comment (chriseth) [reply libyul/optimiser/ReasoningBasedSimplifier.cpp:59]: I would say either combine this commit with the first or also put the creation of `Solver::encodeVariableDeclaration` into this (second) commit.

File: libyul/optimiser/Solver.h [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace solidity::yul
+{
+
+/**
+ * Base class for SMT based optimization steps.
+ *
+ * Works best when used with AST in SSA form.
+ */
+class Solver
```
Comment (leonardoalt) [inline libyul/optimiser/Solver.h]: I would still suggest calling it SMTSolver or something else, as Solver might be confused with the LPSolver or other stuff.

File: libyul/optimiser/Solver.h [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+	smtutil::Expression newVariable();
+	virtual smtutil::Expression newRestrictedVariable(bigint _maxValue = (bigint(1) << 256) - 1);
+	std::string uniqueName();
+
+	virtual std::shared_ptr<smtutil::Sort> defaultSort() const;
+	virtual smtutil::Expression booleanValue(smtutil::Expression _value) const;
+	virtual smtutil::Expression constantValue(bigint _value) const;
+	virtual smtutil::Expression literalValue(Literal const& _literal) const;
+	virtual smtutil::Expression twosComplementToSigned(smtutil::Expression _value);
+	virtual smtutil::Expression signedToTwosComplement(smtutil::Expression _value);
```
Comment (leonardoalt) [inline libyul/optimiser/Solver.h]: The two above can be `const` too

Comment (chriseth) [reply libyul/optimiser/Solver.h]: Interesting, why aren't they static? Does it make sense to override them?

Comment (hrkrshnn) [reply libyul/optimiser/Solver.h]: Yeah, these functions can be both static and non-virtual. What do you think @leonardoalt

Comment (leonardoalt) [reply libyul/optimiser/Solver.h]: Yep makes sense

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**